### DISCUSSION TITLE
Login e Social Login

### DISCUSSION THREAD

**Discussion by j-cvx** [balta-io/7196#40] [Perguntas e Respostas] [answered]
@andrebaltieri eu estou tentando utilizar o código do outro curso da autenticação por token, em conjunto com o Social Login do Firebase deste curso, pra ficar mais próximo de uma solução completa.

No caso, poderia manter os métodos deste curso para a `TodoController` e só mudar a referência para recuperar o usuário (`User.Identity.Name` e `User.Claims.FirstOr`...)? Teria alguma etapa adicional além de adicionar os dois serviços `builder.Services.AddAuthentication` (token e firebase) no Program.cs?

_TodoController_
```
[Route("")]
[HttpGet]
public IActionResult GetAll(
    [FromServices] ITodoItemRepository repository,
    [FromQuery] int page = 0,
    [FromQuery] int pageSize = 25)
{
    try
    {
        // App Login
        var user = User.Identity != null ? User.Identity.Name :
            User.Claims.FirstOrDefault(x => x.Type == "user_id")?.Value;

        var todos = repository.GetAll(user);

        return Ok(new GenericCommandResult(
            true,
            "Todos",
            new { total = todos.Count(), page, pageSize, todos }));
    }
    catch (Exception ex)
    {
        return StatusCode(500, new GenericCommandResult(
            false, "ACC02 - Falha interna do servidor!",
            ex.Message));
    }
}
```

_Program.cs_
```
void ConfigureAuthentication()
{
    var key = Encoding.ASCII.GetBytes(
        builder.Configuration.GetValue<string>("JwtKey"));

    // App Auth
    builder.Services.AddAuthentication(x =>
    {
        x.DefaultAuthenticateScheme = JwtBearerDefaults.AuthenticationScheme;
        x.DefaultChallengeScheme = JwtBearerDefaults.AuthenticationScheme;
    }).AddJwtBearer(x =>
    {
        x.TokenValidationParameters = new TokenValidationParameters
        {
            ValidateIssuerSigningKey = true,
            IssuerSigningKey = new SymmetricSecurityKey(key),
            ValidateIssuer = false,
            ValidateAudience = false
        };
    });

    // Google Auth
    builder.Services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
    .AddJwtBearer(x =>
    {
        x.Authority = "https://securetoken.google.com/<project>";
        x.TokenValidationParameters = new TokenValidationParameters
        {
            ValidateIssuer = true,
            ValidIssuer = "https://securetoken.google.com/<project>",
            ValidateAudience = true,
            ValidAudience = "<project>",
            ValidateLifetime = true
        };
    });
}
```

**Comment by andrebaltieri** [comment, marked as answer]
@j-cvx poderia manter sim, se quiser gerar o token fora e passar pela sua API, sem problemas... mas acho desncessário...

**Reply by j-cvx** [reply]
![Sem título](https://user-images.githubusercontent.com/105296759/173092308-062265fb-7667-4f83-8759-a08d480d410b.png)

Realmente, acaba sendo bem trabalhoso de implementar vários tokens. O meu objetivo seria ter esse tipo de autenticação da imagem (App e Google), mas o token dentro do app parece ser mais útil inicialmente, pois poderia ter outras utilidades, como ser utilizado nos envios de e-mails de confirmação para o usuário, correto? Enfim, esse questionamento seria mais a respeito de desenvolver um app com autenticação "mista", pois essa parece ser uma prática comum no desenvolvimento de apps.

**Reply by andrebaltieri** [reply]
Então, mas foi isto que fizemos no curso... usar autenticação externa... ai recebemos o token e passamos o Token para a API.
A API só valida o token no server do Google por exemplo...

**Reply by j-cvx** [reply]
Certo, eu até consegui implementar os dois _schemas_ de autenticação, modificando o _Default Policy_ para aceitar os dois tipos 

`Services.AddJwtBearer("Auth").AddJwtBearer("Firebase").AddAuthorization(options => options.DefaultPolicy = new AuthorizationPolicyBuilder().AddAuthenticationSchemes("Auth", "Firebase");`

, mas isso acaba gerando muito retrabalho (e dados), o que vai contra o princípio de ter uma autenticação externa.

Mais adiante no curso eu vi que o Firebase também oferece autenticação por e-mail e senha, acaba sendo uma alternativa melhor. Para armazenar configurações extras do usuário (ex. linguagem, timezone, controle de notificações, tema claro/escuro, etc.) eu teria que cadastrar o usuário no banco do app, correto?  Obg pela ajuda, Balta, seus cursos são bem completos e de boa didática.
